File: percolate/src/percolate/utils/chunking.py
"""Token-aware chunking utilities for handling large inputs.

This module provides utilities for splitting large text or record collections
into chunks that fit within model context windows. Uses tiktoken for accurate
token counting and supports both text-based and record-based chunking.

Key features:
- Accurate token counting using tiktoken (with fallback)
- Context window management with overhead accounting
- Record-boundary preservation (never splits records)
- Optimal chunk size calculation per model

Usage:
    # Text chunking
    chunks = chunk_by_tokens(long_text, "claude-sonnet-4-5")

    # Record chunking
    records = [{"id": 1, "data": "..."}, ...]
    chunks = chunk_by_records(records, "claude-sonnet-4-5")
"""

from __future__ import annotations

import json
from typing import Any, Optional

from loguru import logger

# Model context windows (in tokens)
MODEL_CONTEXT_WINDOWS = {
    # Anthropic
    "claude-sonnet-4-5": 200_000,
    "claude-sonnet-4-5-20250929": 200_000,
    "claude-opus-4": 200_000,
    "claude-opus-4-20250514": 200_000,
    "claude-haiku-4-5": 200_000,
    "claude-haiku-4-5-20251001": 200_000,
    # OpenAI
    "gpt-4.1": 128_000,
    "gpt-5": 128_000,
    "gpt-4o": 128_000,
    # Google
    "gemini-3-ultra": 1_000_000,
    "gemini-3-pro": 1_000_000,
    "gemini-3-flash": 1_000_000,
}

# Overhead estimates (in tokens)
DEFAULT_SYSTEM_PROMPT_OVERHEAD = 1500
DEFAULT_OUTPUT_SCHEMA_OVERHEAD = 500
DEFAULT_RESPONSE_BUFFER_RATIO = 0.2
JSON_RECORD_OVERHEAD = 5


def get_optimal_chunk_size(
    model_name: str,
    overhead_tokens: int = DEFAULT_SYSTEM_PROMPT_OVERHEAD + DEFAULT_OUTPUT_SCHEMA_OVERHEAD,
    response_buffer_ratio: float = DEFAULT_RESPONSE_BUFFER_RATIO,
) -> int:
    """Calculate optimal chunk size for a model.

    Accounts for:
    - Model context window
    - System prompt and schema overhead
    - Response buffer space

    Args:
        model_name: LLM model identifier
        overhead_tokens: System prompt + schema overhead (default: 2000)
        response_buffer_ratio: Reserve % for response (default: 20%)

    Returns:
        Maximum tokens per chunk

    Example:
        >>> get_optimal_chunk_size("claude-sonnet-4-5")
        158400  # 200k - 2000 - (200k * 0.2)
    """
    # Get context window for model (default to conservative 100k if unknown)
    context_window = MODEL_CONTEXT_WINDOWS.get(model_name, 100_000)

    # Calculate usable space
    response_buffer = int(context_window * response_buffer_ratio)
    usable_tokens = context_window - overhead_tokens - response_buffer

    logger.debug(
        f"Model {model_name}: context={context_window}, "
        f"overhead={overhead_tokens}, buffer={response_buffer}, "
        f"usable={usable_tokens}"
    )

    return usable_tokens


def estimate_tokens(content: str, model_name: str) -> int:
    """Estimate token count for content.

    Uses tiktoken for accurate counting with fallback to character-based
    estimation if tiktoken is not available.

    Args:
        content: Text to estimate
        model_name: Model for counting

    Returns:
        Estimated token count

    Example:
        >>> estimate_tokens("Hello world", "claude-sonnet-4-5")
        2
    """
    try:
        import tiktoken

        # Map model names to tiktoken encodings
        # Claude models use cl100k_base encoding (same as GPT-4)
        encoding_map = {
            "claude-sonnet-4-5": "cl100k_base",
            "claude-opus-4": "cl100k_base",
            "claude-haiku-4-5": "cl100k_base",
            "gpt-4.1": "cl100k_base",
            "gpt-5": "cl100k_base",
            "gpt-4o": "cl100k_base",
        }

        encoding_name = encoding_map.get(model_name, "cl100k_base")
        encoding = tiktoken.get_encoding(encoding_name)
        return len(encoding.encode(content))
    except ImportError:
        # Fallback to character-based estimation (~4 chars per token)
        logger.warning("tiktoken not available, using character-based estimation")
        return len(content) // 4
    except Exception as e:
        logger.warning(f"Token estimation failed: {e}, using character-based estimation")
        return len(content) // 4


def chunk_by_tokens(
    content: str,
    model_name: str,
    max_chunk_tokens: Optional[int] = None,
) -> list[str]:
    """Split content by token boundaries.

    Uses tiktoken for accurate token counting and splits content into chunks
    that fit within the specified token limit. Attempts to split on sentence
    boundaries where possible to maintain coherence.

    Args:
        content: Text to chunk
        model_name: Model for token counting
        max_chunk_tokens: Override optimal chunk size

    Returns:
        List of text chunks, each within token limit

    Example:
        >>> chunks = chunk_by_tokens(long_text, "claude-sonnet-4-5")
        >>> all(estimate_tokens(c, "claude-sonnet-4-5") <= get_optimal_chunk_size("claude-sonnet-4-5") for c in chunks)
        True
    """
    # Calculate chunk size if not provided
    chunk_size = max_chunk_tokens or get_optimal_chunk_size(model_name)

    # Check if content fits in single chunk
    total_tokens = estimate_tokens(content, model_name)
    if total_tokens <= chunk_size:
        logger.debug(f"Content fits in single chunk ({total_tokens} tokens)")
        return [content]

    logger.info(f"Splitting content ({total_tokens} tokens) into chunks of {chunk_size} tokens")

    chunks = []
    current_chunk = ""
    current_tokens = 0

    # Split on sentences for better coherence
    sentences = _split_into_sentences(content)

    for sentence in sentences:
        sentence_tokens = estimate_tokens(sentence, model_name)

        # If single sentence exceeds chunk size, split it further
        if sentence_tokens > chunk_size:
            # Save current chunk if non-empty
            if current_chunk:
                chunks.append(current_chunk.strip())
                current_chunk = ""
                current_tokens = 0

            # Split long sentence by character chunks
            char_chunks = _split_by_characters(sentence, chunk_size, model_name)
            chunks.extend(char_chunks[:-1])

            # Start new chunk with last piece
            current_chunk = char_chunks[-1]
            current_tokens = estimate_tokens(current_chunk, model_name)

        # Check if adding this sentence would exceed limit
        elif current_tokens + sentence_tokens > chunk_size:
            # Save current chunk and start new one
            chunks.append(current_chunk.strip())
            current_chunk = sentence
            current_tokens = sentence_tokens
        else:
            # Add sentence to current chunk
            current_chunk += (" " if current_chunk else "") + sentence
            current_tokens += sentence_tokens

    # Add final chunk if non-empty
    if current_chunk.strip():
        chunks.append(current_chunk.strip())

    logger.info(f"Created {len(chunks)} chunks")
    return chunks


def _split_into_sentences(text: str) -> list[str]:
    """Split text into sentences on common sentence boundaries.

    Args:
        text: Text to split

    Returns:
        List of sentences
    """
    # Simple sentence splitting on common terminators
    # This is a basic approach - could be enhanced with NLTK or spacy
    import re

    # Split on . ! ? followed by space and capital letter
    sentences = re.split(r'([.!?]+)\s+(?=[A-Z])', text)

    # Recombine punctuation with sentences
    result = []
    for i in range(0, len(sentences) - 1, 2):
        if i + 1 < len(sentences):
            result.append(sentences[i] + sentences[i + 1])
        else:
            result.append(sentences[i])

    # Add last part if exists
    if len(sentences) % 2 == 1:
        result.append(sentences[-1])

    return [s for s in result if s.strip()]


def _split_by_characters(text: str, max_tokens: int, model_name: str) -> list[str]:
    """Split text by character boundaries when sentence splitting isn't enough.

    Args:
        text: Text to split
        max_tokens: Maximum tokens per chunk
        model_name: Model for token counting

    Returns:
        List of text chunks
    """
    # Estimate characters per token (rough approximation)
    avg_chars_per_token = len(text) / estimate_tokens(text, model_name)
    chunk_chars = int(max_tokens * avg_chars_per_token * 0.9)  # 90% to be safe

    chunks = []
    for i in range(0, len(text), chunk_chars):
        chunk = text[i:i + chunk_chars]
        chunks.append(chunk)

    return chunks


def is_list_content(content: str | list) -> bool:
    """Detect if content is a list structure.

    Args:
        content: Content to check

    Returns:
        True if content is a list or JSON array string

    Example:
        >>> is_list_content([{"id": 1}, {"id": 2}])
        True
        >>> is_list_content('[{"id": 1}]')
        True
        >>> is_list_content("plain text")
        False
    """
    if isinstance(content, list):
        return True

    # Try parsing as JSON array
    if isinstance(content, str):
        try:
            parsed = json.loads(content)
            return isinstance(parsed, list)
        except (json.JSONDecodeError, ValueError):
            return False

    return False


def chunk_by_records(
    content: list[dict[str, Any]],
    model_name: str,
    max_records_per_chunk: Optional[int] = None,
) -> list[str]:
    """Chunk list by record boundaries (never splits records).

    Calculates optimal record count based on:
    - Average tokens per record (sampled from first 10)
    - JSON syntax overhead per record
    - Model context window

    Args:
        content: List of records (dicts)
        model_name: Model for token estimation
        max_records_per_chunk: Override calculated optimal

    Returns:
        List of JSON string chunks (complete records only)

    Example:
        >>> records = [{"id": 1, "data": "..."}, {"id": 2, "data": "..."}]
        >>> chunks = chunk_by_records(records, "claude-sonnet-4-5")
        >>> len(chunks) >= 1
        True
    """
    if not content:
        return []

    # Calculate optimal records per chunk
    if max_records_per_chunk is None:
        stats = estimate_record_count(content, model_name)
        max_records_per_chunk = stats["optimal_records_per_chunk"]

    logger.info(f"Chunking {len(content)} records into chunks of {max_records_per_chunk} records")

    # Split into chunks
    chunks = []
    for i in range(0, len(content), max_records_per_chunk):
        chunk_records = content[i:i + max_records_per_chunk]
        chunk_json = json.dumps(chunk_records, indent=2)
        chunks.append(chunk_json)

    logger.info(f"Created {len(chunks)} record chunks")
    return chunks


def estimate_record_count(
    content: list[dict[str, Any]],
    model_name: str,
) -> dict[str, int]:
    """Provide chunking statistics for records.

    Samples first 10 records to estimate average token count per record,
    then calculates optimal chunking parameters.

    Args:
        content: List of records
        model_name: Model for token estimation

    Returns:
        Dict with chunking statistics:
        - total_records: Total number of records
        - total_tokens: Estimated total tokens
        - avg_tokens_per_record: Average tokens per record
        - optimal_records_per_chunk: Recommended records per chunk
        - estimated_chunks: Estimated number of chunks needed

    Example:
        >>> records = [{"id": i, "data": "x" * 100} for i in range(100)]
        >>> stats = estimate_record_count(records, "claude-sonnet-4-5")
        >>> stats["total_records"]
        100
    """
    if not content:
        return {
            "total_records": 0,
            "total_tokens": 0,
            "avg_tokens_per_record": 0,
            "optimal_records_per_chunk": 0,
            "estimated_chunks": 0,
        }

    # Sample first 10 records to estimate average size
    sample_size = min(10, len(content))
    sample = content[:sample_size]
    sample_json = json.dumps(sample, indent=2)
    sample_tokens = estimate_tokens(sample_json, model_name)

    # Calculate average tokens per record (including JSON overhead)
    avg_tokens_per_record = sample_tokens // sample_size + JSON_RECORD_OVERHEAD

    # Calculate total tokens
    total_tokens = avg_tokens_per_record * len(content)

    # Calculate optimal records per chunk
    chunk_size = get_optimal_chunk_size(model_name)
    optimal_records_per_chunk = max(1, chunk_size // avg_tokens_per_record)

    # Calculate estimated chunks
    estimated_chunks = (len(content) + optimal_records_per_chunk - 1) // optimal_records_per_chunk

    stats = {
        "total_records": len(content),
        "total_tokens": total_tokens,
        "avg_tokens_per_record": avg_tokens_per_record,
        "optimal_records_per_chunk": optimal_records_per_chunk,
        "estimated_chunks": estimated_chunks,
    }

    logger.debug(f"Record stats: {stats}")
    return stats

File: percolate-rocks/src/agents/chunking.rs
//! Token-aware content chunking
//!
//! Splits large content into chunks that fit within model context windows.
//! Uses tiktoken-rs for accurate token counting.
//!
//! # Design Philosophy
//!
//! - **Accurate token counting**: Use tiktoken-rs (same as OpenAI/Anthropic)
//! - **Boundary preservation**: Never split mid-sentence or mid-record
//! - **Optimal chunk size**: Calculate based on model context window
//! - **Two modes**: Text chunking (sentences) and record chunking (JSON arrays)
//!
//! # Reference
//!
//! Python implementation: `percolate/src/percolate/utils/chunking.py`
//! This Rust version should match the Python behavior exactly.

/// Token-aware chunker
///
/// Splits content into chunks that fit within a model's context window,
/// preserving natural boundaries (sentences for text, records for JSON).
///
/// # Design Notes
///
/// **Why token-based chunking?**
/// - Character-based chunking is inaccurate (1 char ≠ 1 token)
/// - Word-based chunking ignores tokenization (1 word can be multiple tokens)
/// - Token-based chunking ensures chunks actually fit in context window
///
/// **Context window calculation:**
/// ```text
/// usable_tokens = context_window - overhead - response_buffer
///
/// where:
///   context_window = model's max tokens (e.g., 200k for claude-haiku-4-5)
///   overhead = system prompt + schema (typically ~2k tokens)
///   response_buffer = space for LLM response (20% of window)
/// ```
///
/// **Example for claude-haiku-4-5:**
/// ```text
/// context_window = 200,000
/// overhead = 2,000
/// response_buffer = 40,000 (20%)
/// usable_tokens = 200,000 - 2,000 - 40,000 = 158,000 tokens per chunk
/// ```
///
/// # Example
///
/// ```no_run
/// # use percolate_rocks::agents::Chunker;
/// let chunker = Chunker::new("claude-haiku-4-5", Some(50_000));
/// let large_text = "Very long document...".repeat(1000);
/// let chunks = chunker.chunk_text(&large_text);
///
/// println!("Split into {} chunks", chunks.len());
/// for (i, chunk) in chunks.iter().enumerate() {
///     println!("Chunk {}: {} chars", i, chunk.len());
/// }
/// ```
pub struct Chunker {
    /// Model name for token counting
    ///
    /// Different models use different tokenizers:
    /// - claude-*: cl100k_base (Anthropic uses same as OpenAI)
    /// - gpt-*: cl100k_base (GPT-4) or o200k_base (GPT-4.1)
    pub(crate) model: String,

    /// Maximum tokens per chunk
    ///
    /// If None, calculated automatically based on model context window.
    /// Can be overridden for testing (e.g., force small chunks).
    pub(crate) max_tokens: usize,
}

impl Chunker {
    /// Create a new chunker
    ///
    /// # Arguments
    ///
    /// * `model` - Model name (e.g., "claude-haiku-4-5")
    /// * `max_tokens` - Optional max tokens per chunk (None = auto-calculate)
    ///
    /// # Returns
    ///
    /// Configured chunker ready to split content
    ///
    /// # Implementation Notes (Phase 2)
    ///
    /// **Auto-calculate max_tokens if not provided:**
    /// ```text
    /// 1. Get model context window:
    ///    - claude-haiku-4-5: 200k
    ///    - claude-sonnet-4-5: 200k
    ///    - gpt-4.1: 128k
    ///
    /// 2. Calculate usable tokens:
    ///    overhead = 2000  // System prompt + schema
    ///    response_buffer = context_window * 0.2  // 20% for response
    ///    usable = context_window - overhead - response_buffer
    ///
    /// 3. Set max_tokens = usable
    /// ```
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use percolate_rocks::agents::Chunker;
    /// // Auto-calculate based on model
    /// let chunker1 = Chunker::new("claude-haiku-4-5", None);
    ///
    /// // Override for testing (force small chunks)
    /// let chunker2 = Chunker::new("claude-haiku-4-5", Some(1000));
    /// ```
    pub fn new(model: &str, max_tokens: Option<usize>) -> Self {
        // TODO: Implement in Phase 2
        //
        // Steps:
        // 1. If max_tokens is Some, use it directly
        // 2. If max_tokens is None, calculate from model:
        //    a. Get context window size
        //    b. Subtract overhead (2000 tokens)
        //    c. Subtract response buffer (20%)
        //    d. Use result as max_tokens
        // 3. Store model and max_tokens
        todo!("Create chunker")
    }

    /// Chunk text with sentence boundary preservation
    ///
    /// Splits text into chunks under max_tokens, never breaking mid-sentence.
    ///
    /// # Arguments
    ///
    /// * `content` - Text to chunk
    ///
    /// # Returns
    ///
    /// Vector of text chunks, each under max_tokens
    ///
    /// # Algorithm
    ///
    /// ```text
    /// 1. Estimate total tokens in content
    /// 2. If under max_tokens, return single chunk
    /// 3. Split text into sentences (. ! ? with space after)
    /// 4. Group sentences into chunks:
    ///    - Add sentences to current chunk while under max_tokens
    ///    - When adding next sentence would exceed max_tokens, start new chunk
    /// 5. Return chunks
    /// ```
    ///
    /// # Boundary Preservation
    ///
    /// **Never split mid-sentence:**
    /// ```text
    /// Good:
    ///   Chunk 1: "The cat sat on the mat. The dog ran away."
    ///   Chunk 2: "The bird flew high. The fish swam deep."
    ///
    /// Bad:
    ///   Chunk 1: "The cat sat on the mat. The dog ran aw"
    ///   Chunk 2: "ay. The bird flew high. The fish swam deep."
    /// ```
    ///
    /// # Implementation Notes (Phase 2)
    ///
    /// **Sentence detection (simple approach):**
    /// ```rust
    /// let sentences: Vec<&str> = content
    ///     .split_inclusive(&['.', '!', '?'])
    ///     .filter(|s| !s.trim().is_empty())
    ///     .collect();
    /// ```
    ///
    /// **Token estimation per sentence:**
    /// ```rust
    /// use tiktoken_rs::tokenize;
    ///
    /// for sentence in sentences {
    ///     let tokens = tokenize(&self.model, sentence);
    ///     if current_chunk_tokens + tokens.len() > self.max_tokens {
    ///         // Start new chunk
    ///     } else {
    ///         // Add to current chunk
    ///     }
    /// }
    /// ```
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use percolate_rocks::agents::Chunker;
    /// let chunker = Chunker::new("claude-haiku-4-5", Some(100));
    /// let text = "First sentence. Second sentence. Third sentence. Fourth sentence.";
    /// let chunks = chunker.chunk_text(text);
    ///
    /// // With max_tokens=100, might get 2 chunks:
    /// // chunks[0] = "First sentence. Second sentence."
    /// // chunks[1] = "Third sentence. Fourth sentence."
    /// ```
    pub fn chunk_text(&self, content: &str) -> Vec<String> {
        // TODO: Implement in Phase 2
        //
        // Steps:
        // 1. Estimate total tokens
        // 2. If under max_tokens, return vec![content.to_string()]
        // 3. Split into sentences
        // 4. Group sentences into chunks:
        //    let mut chunks = Vec::new();
        //    let mut current_chunk = String::new();
        //    let mut current_tokens = 0;
        //
        //    for sentence in sentences {
        //        let sentence_tokens = estimate_tokens(sentence);
        //        if current_tokens + sentence_tokens > self.max_tokens {
        //            chunks.push(current_chunk);
        //            current_chunk = sentence.to_string();
        //            current_tokens = sentence_tokens;
        //        } else {
        //            current_chunk.push_str(sentence);
        //            current_tokens += sentence_tokens;
        //        }
        //    }
        //    if !current_chunk.is_empty() {
        //        chunks.push(current_chunk);
        //    }
        // 5. Return chunks
        todo!("Chunk text by tokens with sentence boundaries")
    }

    /// Chunk JSON records preserving record boundaries
    ///
    /// Splits array of JSON objects into chunks, never breaking mid-record.
    ///
    /// # Arguments
    ///
    /// * `records` - Array of JSON objects to chunk
    ///
    /// # Returns
    ///
    /// Vector of record chunks, each under max_tokens
    ///
    /// # Algorithm
    ///
    /// ```text
    /// 1. Sample first 10 records to estimate avg tokens per record
    /// 2. Calculate records_per_chunk = max_tokens / avg_tokens_per_record
    /// 3. Split records into chunks of size records_per_chunk
    /// 4. Return chunks
    /// ```
    ///
    /// # Boundary Preservation
    ///
    /// **Never split mid-record:**
    /// ```text
    /// Good:
    ///   Chunk 1: [{"id": 1}, {"id": 2}]
    ///   Chunk 2: [{"id": 3}, {"id": 4}]
    ///
    /// Bad:
    ///   Chunk 1: [{"id": 1}, {"id": 2}, {"id":]
    ///   Chunk 2: [3}, {"id": 4}]
    /// ```
    ///
    /// # Implementation Notes (Phase 2)
    ///
    /// **Estimate tokens per record:**
    /// ```rust
    /// // Sample first 10 records
    /// let sample_size = records.len().min(10);
    /// let mut total_tokens = 0;
    ///
    /// for record in &records[..sample_size] {
    ///     let json_str = serde_json::to_string(record)?;
    ///     total_tokens += estimate_tokens(&json_str);
    /// }
    ///
    /// let avg_tokens_per_record = total_tokens / sample_size;
    /// let records_per_chunk = self.max_tokens / avg_tokens_per_record;
    /// ```
    ///
    /// **Chunk records:**
    /// ```rust
    /// let chunks: Vec<Vec<serde_json::Value>> = records
    ///     .chunks(records_per_chunk)
    ///     .map(|chunk| chunk.to_vec())
    ///     .collect();
    /// ```
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use percolate_rocks::agents::Chunker;
    /// # use serde_json::json;
    /// let chunker = Chunker::new("claude-haiku-4-5", Some(1000));
    /// let records = vec![
    ///     json!({"id": 1, "name": "Alice"}),
    ///     json!({"id": 2, "name": "Bob"}),
    ///     json!({"id": 3, "name": "Charlie"}),
    /// ];
    /// let chunks = chunker.chunk_records(&records);
    ///
    /// // Might get 2 chunks depending on token size:
    /// // chunks[0] = [{"id": 1, ...}, {"id": 2, ...}]
    /// // chunks[1] = [{"id": 3, ...}]
    /// ```
    pub fn chunk_records(&self, records: &[serde_json::Value]) -> Vec<Vec<serde_json::Value>> {
        // TODO: Implement in Phase 2
        //
        // Steps:
        // 1. If records is empty, return empty vec
        // 2. Sample first 10 records (or fewer if less than 10)
        // 3. Estimate avg tokens per record:
        //    let mut total_tokens = 0;
        //    for record in sample {
        //        let json_str = serde_json::to_string(record)?;
        //        total_tokens += estimate_tokens(&json_str);
        //    }
        //    let avg = total_tokens / sample.len();
        // 4. Calculate records per chunk:
        //    let records_per_chunk = (self.max_tokens / avg).max(1);
        // 5. Split records into chunks:
        //    records.chunks(records_per_chunk)
        //        .map(|chunk| chunk.to_vec())
        //        .collect()
        todo!("Chunk records preserving boundaries")
    }

    /// Estimate token count for content
    ///
    /// Uses tiktoken-rs for accurate counting. Falls back to char count / 4
    /// if tiktoken fails.
    ///
    /// # Arguments
    ///
    /// * `content` - Text to estimate tokens for
    ///
    /// # Returns
    ///
    /// Estimated number of tokens
    ///
    /// # Implementation Notes (Phase 2)
    ///
    /// **Token counting with tiktoken-rs:**
    /// ```rust
    /// use tiktoken_rs::{get_bpe_from_model, tokenize};
    ///
    /// // Get tokenizer for model
    /// let bpe = get_bpe_from_model(&self.model)?;
    ///
    /// // Tokenize content
    /// let tokens = bpe.encode_with_special_tokens(content);
    ///
    /// // Return token count
    /// tokens.len()
    /// ```
    ///
    /// **Fallback if tiktoken unavailable:**
    /// ```rust
    /// // Rough estimate: 1 token ≈ 4 characters
    /// (content.chars().count() / 4).max(1)
    /// ```
    ///
    /// # Accuracy
    ///
    /// **tiktoken-rs is highly accurate** (same tokenizer as OpenAI/Anthropic):
    /// ```text
    /// "Hello world" = 2 tokens
    /// "The quick brown fox" = 4 tokens
    /// "anthropic" = 1 token
    /// "tokenization" = 3 tokens
    /// ```
    ///
    /// **Fallback is approximate** but safe (overestimates):
    /// ```text
    /// "Hello world" (11 chars) → 11/4 = 2 tokens ✓
    /// "The quick brown fox" (19 chars) → 19/4 = 4 tokens ✓
    /// "anthropic" (9 chars) → 9/4 = 2 tokens (actual: 1) ✗
    /// ```
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use percolate_rocks::agents::Chunker;
    /// let chunker = Chunker::new("claude-haiku-4-5", None);
    ///
    /// let short_text = "Hello world";
    /// let tokens = chunker.estimate_tokens(short_text);
    /// assert_eq!(tokens, 2);  // With tiktoken
    ///
    /// let long_text = "word ".repeat(1000);
    /// let tokens = chunker.estimate_tokens(&long_text);
    /// assert!(tokens > 900 && tokens < 1100);  // ~1000 tokens
    /// ```
    pub fn estimate_tokens(&self, content: &str) -> usize {
        // TODO: Implement in Phase 2
        //
        // Steps:
        // 1. Try tiktoken-rs:
        //    use tiktoken_rs::get_bpe_from_model;
        //    if let Ok(bpe) = get_bpe_from_model(&self.model) {
        //        let tokens = bpe.encode_with_special_tokens(content);
        //        return tokens.len();
        //    }
        // 2. Fallback to char count / 4:
        //    (content.chars().count() / 4).max(1)
        todo!("Estimate tokens using tiktoken-rs")
    }

    /// Get model context window size in tokens
    ///
    /// Returns the maximum context window for the model.
    ///
    /// # Returns
    ///
    /// Context window size in tokens
    ///
    /// # Context Windows (2025)
    ///
    /// | Model | Context Window |
    /// |-------|---------------|
    /// | claude-haiku-4-5 | 200,000 |
    /// | claude-sonnet-4-5 | 200,000 |
    /// | claude-opus-4 | 200,000 |
    /// | gpt-4.1 | 128,000 |
    /// | gpt-4.1-turbo | 128,000 |
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use percolate_rocks::agents::Chunker;
    /// let chunker = Chunker::new("claude-haiku-4-5", None);
    /// let window = chunker.get_context_window();
    /// assert_eq!(window, 200_000);
    /// ```
    fn get_context_window(&self) -> usize {
        match self.model.as_str() {
            // Anthropic Claude models
            "claude-haiku-4-5" | "claude-sonnet-4-5" | "claude-opus-4" => 200_000,

            // OpenAI GPT models
            "gpt-4.1" | "gpt-4.1-turbo" => 128_000,

            // Unknown model - conservative default
            _ => {
                tracing::warn!(
                    model = %self.model,
                    "Unknown model context window - using 100k default"
                );
                100_000
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_context_window() {
        let chunker = Chunker {
            model: "claude-haiku-4-5".to_string(),
            max_tokens: 50_000,
        };
        assert_eq!(chunker.get_context_window(), 200_000);

        let chunker = Chunker {
            model: "gpt-4.1".to_string(),
            max_tokens: 50_000,
        };
        assert_eq!(chunker.get_context_window(), 128_000);

        let chunker = Chunker {
            model: "unknown-model".to_string(),
            max_tokens: 50_000,
        };
        assert_eq!(chunker.get_context_window(), 100_000);
    }

    // TODO: Add more tests in Phase 2
    // - test_chunk_text_single_chunk
    // - test_chunk_text_multiple_chunks
    // - test_chunk_text_preserves_sentences
    // - test_chunk_records_single_chunk
    // - test_chunk_records_multiple_chunks
    // - test_estimate_tokens_fallback
}
